etc/bin/swarm

```diff
@@ -0,0 +1,401 @@
+                return 0
+                break
+                ;;
+            [Nn])
+                return 1
+                break
+                ;;
+        esac
+    done
+}
+
+# wrapper of ssh-keygen
+generate_keys () {
+    # the name of the key to generate
+    local KEY="$1"
+
+    if ! [[ -f "$SSHD/$KEY" ]]; then
+        if prompt "(Re)Generate ssh key: $KEY?"; then
+            ssh-keygen \
+                -t rsa \
+                -b 4096 \
+                -C 'docker@localhost' \
+                -f "$SSHD/$KEY" \
+                -P ''
+        else
+            exit 1
+        fi
+    fi
+}
+
+# generate ssh keys for host-container comms
+generate_ssh_keys () {
+    local BASE='.docker-ssh-keys'
+    local KEY
+    mkdir "$BASE" -p
+
+    # ssh key pair for connecting to docker containers
+    KEY='cylc-docker'
+    generate_keys "$KEY" "$BASE"
+    cp "$SSHD/$KEY"* "$BASE"
+}
+
+# add a line to a configuration file touching the file if not present
+append_config () {
+    # the line to add to the config
+    local LINE="$1"
+    # the config to add the line to
+    local LOC="$2"
+    # whether the line should appear at the top or bottom (default bottom)
+    local POS=${3:-bottom}
+
+    # create the config if not already there
+    mkdir -p "$(dirname "$LOC")"
+    touch "$LOC"
+
+    # add the line to the config...
+    if ! grep -q "$LINE" "$LOC"; then
+        if prompt "Write \"$LINE\" to \"$LOC\"?"; then
+            if [[ "$POS" == top ]]; then
+                # ... at the top of the file
+                sed -i "1i$LINE" "$LOC"
+            elif [[ "$POS" == bottom ]]; then 
+                # ... at the bottom of the file
+                echo -e "\n$LINE" >> "$LOC"
+            else
+                # ... nowhere
+                echo "Invalid position :$PWD" >&2
+                exit 1
+            fi
+
+        else
+            exit 1
+        fi
+    fi
+}
+
+# docker run opts for independent filesystem images
+INDEP_FS_OPTS=(
+)
+# docker run opts for shared filesystem images
+SHARED_FS_OPTS=(
+    '-v' "$HOME/cylc-run:/root/cylc-run"
+    #'-p' '43001-43100:43001-43100'  # all cylc run ports
+    '-P'  # all ports
+)
+# docker run opts for tcp communication images
+TCP_OPTS=(
+    '-e' "HOST_HOSTNAME=$(hostname -f)"
+)
+# docker run opts for polling communication images
+POLL_OPTS=(
+    '-e' 'HOST_HOSTNAME=false'
+)
+# the lowest port to use for communicating with containers
+BASE_PORT=42220
+
+# wrapper for running containers with the required options
+_run () {
+    # run a container
+    local fs="$1"  # i.e. indep, shared
+    local comms="$2"  # i.e. tcp, poll
+    local detached="${3:-false}"  # i.e. true, false
+    local port
+
+    # set the detach args
+    local opts=()
+    if ${detached}; then
+        opts+=('-i')
+    else
+        opts+=('-d')
+    fi
+
+    # set the filesystem args
+    if [[ $fs == indep ]]; then
+        port=$BASE_PORT
+        opts+=("${INDEP_FS_OPTS[@]}")
```

Comment:
`"${EMPTY_ARRAY[@]}"` is undefined and fails `set -u` prior to bash-4.4 (and my Centos environment is still bash-4.2, damn it).